### PR TITLE
Add Gen module and propery tests

### PR DESCRIPTION
**Pull request by safareli**

### CODE REVIEW

File: src/Data/Path/Pathy/Gen.purs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import Control.Monad.Gen as Gen
+import Control.Monad.Rec.Class (class MonadRec)
+import Data.Either (Either(..))
+import Data.Foldable (foldr)
+import Data.List as L
+import Data.NonEmpty ((:|))
+import Data.Path.Pathy (AbsPath, AbsFile, AbsDir, RelDir, RelFile, RelPath, Sandboxed, (</>))
+import Data.Path.Pathy as P
+
+genAbsDirPath :: forall m. MonadGen m => MonadRec m => m String -> m (AbsDir Sandboxed)
```
Comment (safareli) [inline src/Data/Path/Pathy/Gen.purs]: This staff is moved from ps-quasar

File: test/Main.purs [hunk 2]
```diff
@@ -39,8 +40,37 @@ instance arbitraryArbPath ∷ QC.Arbitrary ArbPath where
 pathPart ∷ Gen.Gen String
 pathPart = Gen.suchThat QC.arbitrary (not <<< Str.null)
 
+parsePrintCheck :: forall a b. Path a b Sandboxed -> Maybe (Path a b Unsandboxed) -> QC.Result
+parsePrintCheck input parsed =
+  if parsed == Just (unsandbox input)
```
Comment (safareli) [inline test/Main.purs:45]: I guess this check is not the best, as I got this error at some point:
```
checking `parse <<< print` for `RelFile``
97/100 test(s) passed.
/Users/safareli/dev/slamdata/purescript-pathy/output/Control.Monad.Eff.Exception/foreign.js:29
    throw e;
    ^

Error: Test 40 (seed 846546249) failed: 
`parse (print path) != Just path` for path: `(((((((((currentDir </> dir "냫笼嬋ఱ护⿼") </> dir "ꘓၷ两뱖쌜") </> dir "沌뢑朿ガ茧") </> dir "ﹱ捴ኧ嵹") </> dir "횽") </> dir "⤋") </> dir "큛रꃙ") </> dir "/솛鼒") </> file "냫笼嬋ఱ护⿼")` which was re-parsed into `(Just (((((((((currentDir </> dir "냫笼嬋ఱ护⿼") </> dir "ꘓၷ两뱖쌜") </> dir "沌뢑朿ガ茧") </> dir "ﹱ捴ኧ嵹") </> dir "횽") </> dir "⤋") </> dir "큛रꃙ") </> dir "솛鼒") </> file "냫笼嬋ఱ护⿼"))`
	Printed path: "./냫笼嬋ఱ护⿼/ꘓၷ两뱖쌜/沌뢑朿ガ茧/ﹱ捴ኧ嵹/횽/⤋/큛रꃙ/솛鼒/냫笼嬋ఱ护⿼"
	Printed path': `(Just "./냫笼嬋ఱ护⿼/ꘓၷ两뱖쌜/沌뢑朿ガ茧/ﹱ捴ኧ嵹/횽/⤋/큛रꃙ/솛鼒/냫笼嬋ఱ护⿼")`
```

if we move show-ed values side by side difference becomes more clear (scroll to right end)
```
(((((((((currentDir </> dir "냫笼嬋ఱ护⿼") </> dir "ꘓၷ两뱖쌜") </> dir "沌뢑朿ガ茧") </> dir "ﹱ捴ኧ嵹") </> dir "횽") </> dir "⤋") </> dir "큛रꃙ") </> dir "/솛鼒") </> file "냫笼嬋ఱ护⿼")
(((((((((currentDir </> dir "냫笼嬋ఱ护⿼") </> dir "ꘓၷ两뱖쌜") </> dir "沌뢑朿ガ茧") </> dir "ﹱ捴ኧ嵹") </> dir "횽") </> dir "⤋") </> dir "큛रꃙ") </> dir "솛鼒") </> file "냫笼嬋ఱ护⿼")
```

The different part:
```
... </> dir "/솛鼒") </> file "냫笼嬋ఱ护⿼")
... </> dir "솛鼒") </> file "냫笼嬋ఱ护⿼")
```

Comment (safareli) [reply test/Main.purs:45]: Used alpha numeric generator in resent amend so this issue is not going to occur

File: src/Data/Path/Pathy/Gen.purs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import Data.Either (Either(..))
+import Data.Foldable (foldr)
+import Data.List as L
+import Data.NonEmpty ((:|))
+import Data.Path.Pathy (AbsPath, AbsFile, AbsDir, RelDir, RelFile, RelPath, Sandboxed, (</>))
+import Data.Path.Pathy as P
+import Data.String.Gen as SG
+
+genName ∷ ∀ m. MonadGen m ⇒ MonadRec m ⇒ m String
+genName = SG.genString $ Gen.oneOf $ CG.genDigitChar :| [CG.genAlpha]
```
Comment (cryogenian) [inline src/Data/Path/Pathy/Gen.purs:25]: https://github.com/purescript-contrib/purescript-strongcheck/blob/master/src/Test/StrongCheck/Data/AlphaNumString.purs#L10

**Review by cryogenian — COMMENTED** [review]
There is a module that produces alphaNumString

File: src/Data/Path/Pathy/Gen.purs (hunk 3 above)
Comment (safareli) [reply src/Data/Path/Pathy/Gen.purs:25]: But we have no dependency on that.

**Comment by safareli** [comment]
@cryogenian is this fine to merge?

**Comment by cryogenian** [comment]
Oh, we don't have strongcheck here, weird... 
Yeah, it's good to go :+1:

**Review by cryogenian — APPROVED** [review]

**Comment by safareli** [comment]
made 4.1.0 release

**Pull request merged**